core/src/main/java/org/testcontainers/utility/ContainerReaper.java

```diff
@@ -0,0 +1,80 @@
+package org.testcontainers.utility;
+
+import com.github.dockerjava.api.DockerClient;
+import com.github.dockerjava.api.DockerException;
+import com.github.dockerjava.api.InternalServerErrorException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.DockerClientFactory;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Component that responsible for container removal and automatic cleanup of dead containers at JVM shutdown.
+ */
+public final class ContainerReaper {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ContainerReaper.class);
+    private static ContainerReaper instance;
+    private final DockerClient dockerClient;
+    private Map<String, String> registeredContainers = new ConcurrentHashMap<>();
+
+    private ContainerReaper() {
+        dockerClient = DockerClientFactory.instance().client();
+
+        // If the JVM stops without containers being stopped, try and stop the container.
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            for (Map.Entry<String, String> entry : new HashSet<>(registeredContainers.entrySet())) {
```

Comment:
Yep sure - as is the code has also to get around a ConcurrentModificationException (removal from the map in `stopAndRemoveContainer`), hence the clunky copy into a new HashSet.

I don't think forEach would be immune. However, I think I'll approach the removal from the map in a cleaner way...